File: morpheus/_lib/src/messages/memory/tensor_memory.cpp
/*
 * SPDX-FileCopyrightText: Copyright (c) 2021-2023, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
 * SPDX-License-Identifier: Apache-2.0
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "morpheus/messages/memory/tensor_memory.hpp"  // IWYU pragma: associated

#include "morpheus/objects/tensor_object.hpp"  // for TensorObject
#include "morpheus/utilities/cupy_util.hpp"    // for CupyUtil
#include "morpheus/utilities/stage_util.hpp"
#include "morpheus/utilities/string_util.hpp"  // for MORPHEUS_CONCAT_STR

#include <pybind11/cast.h>
#include <pybind11/pytypes.h>  // for attribute_error, key_error
#include <pybind11/stl.h>      // IWYU pragma: keep

#include <map>
#include <sstream>    // needed by MORPHEUS_CONCAT_STR
#include <stdexcept>  // for std::length_error
#include <string>
#include <utility>  // for move, pair
#include <vector>
// IWYU pragma: no_include <type_traits>

namespace morpheus {
/****** Component public implementations *******************/
/****** TensorMemory****************************************/
TensorMemory::TensorMemory(TensorIndex count) : count(count) {}
TensorMemory::TensorMemory(TensorIndex count, TensorMap&& tensors) : count(count), m_tensors(std::move(tensors))
{
    check_tensors_length(this->m_tensors);
}

bool TensorMemory::has_tensor(const std::string& name) const
{
    return this->m_tensors.find(name) != this->m_tensors.end();
}

const TensorObject& TensorMemory::get_tensor(const std::string& name) const
{
    verify_tensor_exists(name);
    return m_tensors.at(name);
}

TensorObject& TensorMemory::get_tensor(const std::string& name)
{
    verify_tensor_exists(name);
    return m_tensors[name];
}

void TensorMemory::set_tensor(const std::string& name, TensorObject&& tensor)
{
    check_tensor_length(tensor);
    this->m_tensors[name].swap(std::move(tensor));
}

const TensorMap& TensorMemory::get_tensors() const
{
    return m_tensors;
}

void TensorMemory::set_tensors(TensorMap&& tensors)
{
    check_tensors_length(tensors);
    this->m_tensors = std::move(tensors);
}

TensorMap TensorMemory::copy_tensor_ranges(const std::vector<RangeType>& ranges, TensorIndex num_selected_rows) const
{
    TensorMap tensors;
    for (const auto& p : this->m_tensors)
    {
        tensors.insert(std::pair{p.first, p.second.copy_rows(ranges, num_selected_rows)});
    }

    return tensors;
}

void TensorMemory::check_tensor_length(const TensorObject& tensor)
{
    if (tensor.shape(0) != this->count)
    {
        throw std::length_error{MORPHEUS_CONCAT_STR("The number rows in tensor "
                                                    << tensor.shape(0) << " does not match TensorMemory.count of "
                                                    << this->count)};
    }
}

void TensorMemory::check_tensors_length(const TensorMap& tensors)
{
    for (const auto& p : tensors)
    {
        check_tensor_length(p.second);
    }
}

void TensorMemory::verify_tensor_exists(const std::string& name) const
{
    if (!has_tensor(name))
    {
        throw std::runtime_error(MORPHEUS_CONCAT_STR("Tensor: '" << name << "' not found in memory"));
    }
}

/****** TensorMemoryInterfaceProxy *************************/
std::shared_ptr<TensorMemory> TensorMemoryInterfaceProxy::init(TensorIndex count, pybind11::object& tensors)
{
    if (tensors.is_none())
    {
        return std::make_shared<TensorMemory>(count);
    }
    else
    {
        return std::make_shared<TensorMemory>(
            count, std::move(CupyUtil::cupy_to_tensors(tensors.cast<CupyUtil::py_tensor_map_t>())));
    }
}

TensorIndex TensorMemoryInterfaceProxy::get_count(TensorMemory& self)
{
    return self.count;
}

std::vector<std::string> TensorMemoryInterfaceProxy::tensor_names_getter(TensorMemory& self)
{
    return foreach_map(self.get_tensors(), [](const auto& item) -> std::string {
        // Just return the keys
        return item.first;
    });
}

bool TensorMemoryInterfaceProxy::has_tensor(TensorMemory& self, std::string name)
{
    return self.has_tensor(name);
}

CupyUtil::py_tensor_map_t TensorMemoryInterfaceProxy::get_tensors(TensorMemory& self)
{
    return CupyUtil::tensors_to_cupy(self.get_tensors());
}

void TensorMemoryInterfaceProxy::set_tensors(TensorMemory& self, CupyUtil::py_tensor_map_t tensors)
{
    self.set_tensors(CupyUtil::cupy_to_tensors(tensors));
}

pybind11::object TensorMemoryInterfaceProxy::get_tensor(TensorMemory& self, const std::string name)
{
    try
    {
        auto tensor = self.get_tensor(name);
        return CupyUtil::tensor_to_cupy(tensor);
    } catch (const std::runtime_error& e)
    {
        throw pybind11::key_error{e.what()};
    }
}

pybind11::object TensorMemoryInterfaceProxy::get_tensor_property(TensorMemory& self, const std::string name)
{
    try
    {
        return get_tensor(self, std::move(name));
    } catch (const pybind11::key_error& e)
    {
        throw pybind11::attribute_error{e.what()};
    }
}

void TensorMemoryInterfaceProxy::set_tensor(TensorMemory& self,
                                            const std::string name,
                                            const pybind11::object& cupy_tensor)
{
    self.set_tensor(name, CupyUtil::cupy_to_tensor(cupy_tensor));
}

}  // namespace morpheus

File: morpheus/messages/tensor_memory.py
# SPDX-FileCopyrightText: Copyright (c) 2022-2023, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
# SPDX-License-Identifier: Apache-2.0
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import dataclasses
import typing

import cupy as cp

import morpheus._lib.messages as _messages
from morpheus.messages.message_base import MessageData


@dataclasses.dataclass(init=False)
class TensorMemory(MessageData, cpp_class=_messages.TensorMemory):
    """
    This is a base container class for data that will be used for inference stages. This class is designed to
    hold generic tensor data in cupy arrays.

    Parameters
    ----------
    count : int
        Length of each tensor contained in `tensors`.
    tensors : typing.Dict[str, cupy.ndarray]
        Collection of tensors uniquely identified by a name.

    """
    count: int

    def __init__(self, count: int, tensors: typing.Dict[str, cp.ndarray] = None):
        self.count = count

        if tensors is None:
            tensors = {}
        else:
            self._check_tensors(tensors)

        self._tensors = tensors

    def _check_tensors(self, tensors: typing.Dict[str, cp.ndarray]):
        for tensor in tensors.values():
            self._check_tensor(tensor)

    def _check_tensor(self, tensor: cp.ndarray):
        if (tensor.shape[0] != self.count):
            class_name = type(self).__name__
            raise ValueError(
                f"The number rows in tensor {tensor.shape[0]} does not match {class_name}.count of {self.count}")

    def get_tensors(self):
        """
        Get the tensors contained by this instance. It is important to note that when C++ execution is enabled the
        returned tensors will be a Python copy of the tensors stored in the C++ object. As such any changes made to the
        tensors will need to be updated with a call to `set_tensors`.

        Returns
        -------
        typing.Dict[str, cp.ndarray]
        """
        return self._tensors

    def set_tensors(self, tensors):
        """
        Overwrite the tensors stored by this instance. If the length of the tensors has changed, then the `count`
        property should also be updated.

        Parameters
        ----------
        tensors : typing.Dict[str, cupy.ndarray]
            Collection of tensors uniquely identified by a name.
        """
        self._check_tensors(tensors)
        self._tensors = tensors

    def get_tensor(self, name):
        """
        Get the Tensor stored in the TensorMemory container identified by `name`.

        Parameters
        ----------
        name : str
            Tensor key name.

        Returns
        -------
        cupy.ndarray
            Tensor.

        Raises
        ------
        KeyError
            When no matching tensor exists.
        """
        return self._tensors[name]

    def set_tensor(self, name, tensor):
        """
        Update the tensor identified by `name`. If the length of the tensor has changed, then the `count`
        property should also be updated.

        Parameters
        ----------
        tensors : typing.Dict[str, cupy.ndarray]
            Collection of tensors uniquely identified by a name.
        """
        self._check_tensor(tensor)
        self._tensors[name] = tensor
